### PR TITLE
wasm support and zig-0.12.0 fixes

### PR DESCRIPTION
**Pull request by kassane**
## Tested
- OS: Linux (archlinux)
- Arch: x86_64 (zen3)
- zig version: `0.12.0-dev.2150+63de8a598`

### Steps

- zig get emsdk using `zon` file.
- run emsdk (download, extracts files)
- build libsokol w/ emscripten (os-tag)
- in examples run `emcc` and `emrun`

**Note:** This is experimental, and some examples get black screen on browser or get errors!!

### Reference
- https://github.com/floooh/sokol-zig/issues/8#issuecomment-1877562076

### CODE REVIEW

**Comment by kassane** [comment]
wow! I'll analyze the suggestion and adjust with the new zig update which breaks the previous build.
- https://github.com/ziglang/zig/commit/142471fcc46070326526e3976f0150fe734df0b6

**Comment by kassane** [comment]
My suggestion:

**from:**
https://github.com/floooh/sokol-zig/blob/2a0d82e1716974d4ff6f9e8a4bf8622b50e5704a/build.zig#L18-L26
https://github.com/floooh/sokol-zig/blob/2a0d82e1716974d4ff6f9e8a4bf8622b50e5704a/build.zig#L106-L109
**to:**
https://github.com/floooh/sokol-zig/blob/e176d2b6bc79b5b13b7fbf7822aec958ba1c2792/build.zig#L19-L34
https://github.com/floooh/sokol-zig/blob/e176d2b6bc79b5b13b7fbf7822aec958ba1c2792/build.zig#L231-L237

**Comment by kassane** [comment]
Current issue:

No error:
https://github.com/floooh/sokol-zig/blob/f5598ec4957b2546981e26b4acea170badf83e7b/build.zig#L346

```bash
warning: undefined symbol: sgl_begin_quads (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_context_draw (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_context_make_pipeline (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_default_context (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_defaults (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_enable_texture (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_end (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_load_pipeline (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_lookat (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_make_context (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_matrix_mode_modelview (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_matrix_mode_projection (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_perspective (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_rad (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_rotate (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_set_context (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_setup (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_shutdown (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_texture (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_v2f_c3b (referenced by root reference (e.g. compiled C/C++ code))
warning: undefined symbol: sgl_v3f_t2f (referenced by root reference (e.g. compiled C/C++ code))
emcc: warning: warnings in JS library compilation [-Wjs-compiler]
```

**Comment by floooh** [comment]
Phew, FYI I just fixed build.zig for the latest build system changes in the nightly version (in a new branch zig-0.12.0):

https://github.com/floooh/sokol-zig/tree/zig-0.12.0

I started this branch with the sokol-package branch, so that at least there won't be two version (package and non-package) from there on.

This is getting a bit messy with the wasm changes on top, I hope the build system APIs stabilize a bit better soon-ish.

**Comment by kassane** [comment]
> Phew, FYI I just fixed build.zig for the latest build system changes in the nightly version (in a new branch zig-0.12.0):
> 
> https://github.com/floooh/sokol-zig/tree/zig-0.12.0
> 
> I started this branch with the sokol-package branch, so that at least there won't be two version (package and non-package) from there on.

Amazing!! :+1: 

> This is getting a bit messy with the wasm changes on top, I hope the build system APIs stabilize a bit better soon-ish.

I also hope! Because right now, I was looking into a new issue in **sokol-d** for **windows-msvc** target. Since the latest zig update `0.12.0-dev.2043+6ebeb85ab`.

**Comment by kassane** [comment]
Rebased on **branch:** `package`

**Comment by floooh** [comment]
FYI I think I will spend the next couple of days with a couple of fixes:

I had started a zig-0.12.0 branch in sokol-zig which has been updated to the latest zig build system API changes (I'm not happy with some changes I did, and will use some of changes which make more sense, like `lib.target.isDarwin()` => `lib.rootModuleTarget().isDarwin()` (I hadn't seen this new rootModuleTarget() helper).

I think I will just merge the current master branch back into zig-0.11.0 and treat this as the stable target for people using zig 0.11.x, and instead merge zig-0.12.0 into master (meaning the sokol-zig master branch would be "bleeding edge").

This also means that the sokol-zig master becomes 'package-capable', and I will move the master branches of my example/test projects pacman.zig and kc85.zig also to use the package manager.

I'll keep working on the zig-0.12.0 branch until everything is ready. I'll only do minimal changes to the already existing wasm support and try not to interfere too much with your wip changes :)

**Comment by floooh** [comment]
PPS: One question that popped into my mind... I wonder if it makes sense to split the examples into a separate build.zig (maybe in a subdirectory). For instance adding Emscripten/WASM build support for the samples adds a lot more stuff to the root build.zig which wouldn't be needed for the sokol package alone.

Thoughts?

**Comment by kassane** [comment]
> I had started a zig-0.12.0 branch in sokol-zig which has been updated to the latest zig build system API changes

Redesigning the new api is pretty tedious, besides looking more verbose.
However, the real difficulty is knowing how the zig will handle the wasm target.
The use of emcc is minimized when a wasm executable is generated by zig. This is because the zig compiler itself will have to do the linking, exporting, and relocating of functions.

Thoughts?

> I wonder if it makes sense to split the examples into a separate build.zig (maybe in a subdirectory). For instance adding Emscripten/WASM build support for the samples adds a lot more stuff to the root build.zig which wouldn't be needed for the sokol package alone.

That could be! However, this reorganization of build module would require a more complete configuration of emscripten.

Something similar here, perhaps:
- https://github.com/zig-gamedev/zig-gamedev/pull/309

**Comment by floooh** [comment]
> The use of emcc is minimized when a wasm executable is generated by zig. This is because the zig compiler itself will have to do the linking, exporting, and relocating of functions.

In case of sokol, emcc will need to do the linking, because the sokol C sources use Emscripten-specific magic like embedded Javascript snippets via EM_JS() which need special handling in the linker step. Also the output must not just be a .wasm file, but also a .html and .js file.

For just compiling the sokol static link library, the Zig compiler is sufficient (with the `wasm32-emscripten` target), but for linking additional stuff needs to happen which is only implemented in emcc.

So one way or another, if we want to have the sokol-zig examples built out of the box, we need to be able calling `emcc`.

To just build the sokol library, only a path to the Emscripten SDK sysroot is needed to provide the Emscripten C headers.

**Comment by floooh** [comment]
Btw I'm currently updating pacman.zig to make the WASM support work again and two thoughts:

- first: there may be some breakage or some build system workarounds don't work anymore (for instance it looks like the sokol C library is not compiled with the `wasm32-emscripten` target anymore, even though it should be patched in the sokol-zig build file, will need to investigate
- second, and more importantly: I forgot about the amount of hacks I had to add to Pacman.zig in order to make it work with the Emscripten linker. For instance the Zig main() function can't be used, instead a separate C entry file is needed which has a C main() function (see here: https://github.com/floooh/pacman.zig/tree/main/src/emscripten). Currently I'm also getting an error from wasm-ld that there are two main function, not sure yet if this is because of Emscripten or the latest Zig version.

Long story short: because of all the hacks and workarounds, I don't think anymore to make the sokol-zig samples compile out of the box for WASM/Emscripten. It's better to have an example in pacman.zig and use that as "guinea pig" for the WASM build target.

I'll let you know when I have WASM/Emscripten working in Pacman.zig again (might be a few days)

**Comment by floooh** [comment]
> [kassane](https://github.com/kassane) marked this pull request as ready for review [1 minute ago](https://github.com/floooh/sokol-zig/pull/50#event-11446651226)

...wait... did you actually get it to work?

**Comment by floooh** [comment]
This is pretty awesome tbh, didn't know that this works (but hoped it would):

```
.{
    .name = "sokol-zig",
    .version = "0.0.0",
    .minimum_zig_version = "0.12.0",
    .dependencies = .{
        .emsdk = .{
            .url = "git+https://github.com/emscripten-core/emsdk#0bbae74935d57ff41739648c12cf90b56668398f",
            .hash = "1220f1340cd871b444021c600661f921f96091ce0815fa43008528f4844cece7e245",
        },
    },
    .paths = .{""},
}
```

...means I can probably do something similar with the sokol-shc compiler...

**Comment by floooh** [comment]
woah, what sorcery is this :D

image

**Comment by kassane** [comment]
> woah, what sorcery is this :D

https://github.com/floooh/sokol-zig/blob/ae79928797b0b17671574e13a1a29dd96c1d889a/build.zig#L398-L400

**Comment by kassane** [comment]
All tests :heavy_check_mark: \o/

**Comment by floooh** [comment]
This is totally awesome....

The clear-sample works (which means all the Emscripten stuff is working).

image

...for the samples which require shaders the GLES3/WebGL2 shaders seem to be missing (which figures because they are not currently in the generated shader files here: https://github.com/floooh/sokol-zig/tree/master/src/examples/shaders, and the args here:

https://github.com/floooh/sokol-zig/blob/24134d507a9b6d61ea1c438c0ec2ffd19530a93c/build.zig#L248

...need to be extended like this:

`glsl330:metal_macos:hlsl4:glsl300es`

(maybe also `:wgpu` but let's not get ahead of ourselves...)

We should also replace that ugly Emscripten default webpage which our own shell.html... but that's also no show stopper for the merge, I can take care of this polishing stuff...

In any case, I will abandon my wip work on pacman.zig and kc85.zig and focus on this MR first.

Good stuff :)

**Comment by floooh** [comment]
...not today though (but expect some potential review feedback bits and pieces over the next few days).

**Comment by kassane** [comment]
> This is pretty awesome tbh, didn't know that this works (but hoped it would):
> 
> ```
> .{
>     .name = "sokol-zig",
>     .version = "0.0.0",
>     .minimum_zig_version = "0.12.0",
>     .dependencies = .{
>         .emsdk = .{
>             .url = "git+https://github.com/emscripten-core/emsdk#0bbae74935d57ff41739648c12cf90b56668398f",
>             .hash = "1220f1340cd871b444021c600661f921f96091ce0815fa43008528f4844cece7e245",
>         },
>     },
>     .paths = .{""},
> }
> ```
> 

You can update the package using the command:
```bash
# (over)write [zon] with new commit hash pkg
>$ zig fetch --save=emsdk "git+https://github.com/emscripten-core/emsdk#{latest-commit-hash}"
```

> ...means I can probably do something similar with the `sokol-shc` compiler...

Yeah! I was also thinking about this possibility in the sokol-d project.

**Comment by floooh** [comment]
Hmm strange, I just updated the package branch with the latest changes from master, but the PR diff view still shows differences in sokol_app.h, sokol_gfx.h and sokol_log.h which shouldn't be there anymore...

**Comment by floooh** [comment]
...changing base branches to update the view didn't work though hmm...

File: build.zig [hunk 1]
```diff
@@ -162,42 +202,42 @@ pub fn buildLibSokol(b: *Build, options: LibSokolOptions) !*CompileStep {
...
-        } else if (lib.target.getOsTag() == .macos) {
+        } else if (lib.rootModuleTarget().os.tag == .macos) {
             lib.linkFramework("Cocoa");
             lib.linkFramework("QuartzCore");
             if (.gl == _backend) {
                 lib.linkFramework("OpenGL");
             }
         }
-    } else if (lib.target.getAbi() == .android) {
+    } else if (lib.rootModuleTarget().abi == .android) {
```
Comment (floooh) [inline build.zig:219]: I think there's an `isAndroid()` method now, might be the better choice because this is the only branch which checks for 'abi' instead of 'os.tag'.

File: build.zig [hunk 2]
```diff
@@ -301,3 +351,59 @@ fn buildShaders(b: *Build) void {
...
+    const emcc_path = b.findProgram(&.{"emcc"}, &.{}) catch b.pathJoin(&.{ options.packagePath(b).?, "upstream", "emscripten", "emcc" });
+    const emrun_path = b.findProgram(&.{"emrun"}, &.{}) catch b.pathJoin(&.{ options.packagePath(b).?, "upstream", "emscripten", "emrun" });
+
+    if (options.lib_sokol.rootModuleTarget().os.tag != .emscripten) {
+        std.log.err("Please build with 'zig build -Dtarget=wasm32-emscripten", .{});
+        return error.Wasm32EmscriptenExpected;
+    }
+
+    // call the emcc linker step as a 'system command' zig build step which
+    // depends on the libsokol and libgame build steps
```
Comment (floooh) [inline build.zig:366]: The `libgame` in the comment is a copy-paste-artifact :) I think it's better to rewrite the comment so that it still explains that we need to call the `emcc` linker but remove the outdated information.

File: build.zig [hunk 3]
```diff
@@ -301,3 +351,59 @@ fn buildShaders(b: *Build) void {
...
+    try std.fs.cwd().makePath(b.fmt("{s}/web", .{b.install_path}));
+
+    var emcc_cmds = std.ArrayList([]const u8).init(b.allocator);
```
Comment (floooh) [inline build.zig:369]: Probably better to name this `emcc_cmd`, since it's just a single command (even though it's composed from multiple args). Sorry for nitpicking ;)

File: build.zig [hunk 4]
```diff
@@ -301,3 +351,59 @@ fn buildShaders(b: *Build) void {
...
+    // TODO: fix undefined references
+    // switch (options.backend) {
+    //     .wgpu => {
+    // try emcc_cmds.append("-sUSE_WEBGPU=1");
+    // },
+    // else => {
+    try emcc_cmds.append("-sUSE_WEBGL2=1");
+    //     },
+    // }
+    try emcc_cmds.append("-sEXPORTED_FUNCTIONS=['_malloc','_free','_main']");
```
Comment (floooh) [inline build.zig:394]: It looks like this `-sEXPORTED_FUNCTIONS` stuff is also no longer needed (not sure why tbh, but I just deleted it locally and the samples still build), and the clear sample still runs (and that definitely allocates).

**Comment by floooh** [comment]
I added a couple of code review comments / change requests.

Two things are still mysterious to me:

- why doesn't Github pick up the update to the `package` base branch (because there shouldn't be any differences in the C headers anymore)

- why isn't this `main()` stub hack needed anymore like here in pacman.zig: https://github.com/floooh/pacman.zig/blob/main/src/emscripten/entry.c (my guess is that this was also fixed as part of https://github.com/ziglang/zig/issues/10836

File: build.zig [hunk 5]
```diff
@@ -83,16 +92,24 @@ const ExampleOptions = struct {
     optimize: OptimizeMode,
     lib_sokol: *CompileStep,
     mod_sokol: *Module,
+    package: ?*Build.Dependency = null,
+
+    fn packagePath(self: ExampleOptions, b: *Build) ?[]const u8 {
+        if (self.package) |dep|
+            return dep.path("").getPath(b);
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (kassane) [reply build.zig:101]: I have left the name generic. This is in case it is reused in another package. It's not only private to emsdk.
I should also change the local emsdk_path in the build_wasm function.

File: build.zig [hunk 6]
```diff
@@ -102,19 +119,42 @@ pub fn buildLibSokol(b: *Build, options: LibSokolOptions) !*CompileStep {
         .link_libc = true,
     });
     if (is_wasm) {
-        // need to add Emscripten SDK include path
-        if (b.sysroot == null) {
+        // need to add Emscripten SDK include path (system or package)
+        if (b.sysroot) |sysroot| {
+            config.sysroot = sysroot;
+        } else if (options.packagePath(b)) |path| {
+            config.sysroot = b.pathJoin(&.{ path, "upstream", "emscripten", "cache", "sysroot" });
```
[the remark this answers is not in the document]
Comment (kassane) [reply build.zig:126]: Ditto https://github.com/floooh/sokol-zig/pull/50#discussion_r1449203313

File: build.zig [hunk 7]
```diff
@@ -102,19 +119,42 @@ pub fn buildLibSokol(b: *Build, options: LibSokolOptions) !*CompileStep {
...
+                try cmds.append(b.pathJoin(&.{ options.packagePath(b).?, "emsdk.bat" }))
+            else {
+                try cmds.append("bash"); // or try chmod
+                try cmds.append(b.pathJoin(&.{ options.packagePath(b).?, "emsdk" }));
+            }
+
+            var emsdk_run = b.addSystemCommand(cmds.items);
+            emsdk_run.addArgs(&.{ "install", "latest" });
+            var emsdk_active = b.addSystemCommand(cmds.items);
+            emsdk_active.addArgs(&.{ "activate", "latest" });
```
[the remark this answers is not in the document]
Comment (kassane) [reply build.zig:134]: `std.fs.Dir.stat` to check file or dir exists?

[the remark this answers is not in the document]
Comment (kassane) [reply build.zig:134]: I tried to scan files and was not successful. Also, I'm not sure how to cache the external commands after the changes. But you can suppress emsdk_env output. https://github.com/floooh/sokol-zig/pull/50/commits/650476d1a96d0557cef1b735200c8ab666af9c3d

**Comment by kassane** [comment]
> ...changing base branches to update the view didn't work though hmm...

Do you want to rebase to branch `0.12.0`?
Can I also revert the .c files? Just keep build.zig for changes.

**Comment by floooh** [comment]
> Do you want to rebase to branch 0.12.0?

That was my plan yesterday, but there will be collisions with the changes I already did there (and those changes don't make sense anymore, I would like to use your branch as the "0.12.0 reference branch").

In general my plan is:

- change the 'branch philosophy' of sokol-zig so that the main branch works with the latest Zig *dev* version, instead of the stable version (since most Zig users currently seem to work on the dev version anyway)
- ...and instead turn the 0.11.0, 0.10.0 etc... branches into "archival branches" which keep working with that version of Zig, but might not have the latest sokol headers.

...e.g. the main branch becomes the 'bleeding edge'.

Now I already started a 0.12.0 branch prematurely. It would have been better to merge your changes back into the package branch and then merge the package branch back into main, and only create a 0.12.0 branch when Zig 0.12.0 has been released.

TL;DR: keep this PR based on the `package` branch, and I will figure something out :D

I'll check out your changes later today. After merging back into `package` I will probably fix the remaining issues in the samples that currently produce a black screen, then update my pacman.zig and kc85.zig projects accordingly and then merge everything back into master/main.

And I think I will also generally switch everything to using packages, and declare the old way of integrating the bindings as 'deprecated' (but probably still keep it working for a while).

**Comment by floooh** [comment]
> Can I also revert the .c files? Just keep build.zig for changes.

Yes, ideally the PR would only affect the main.yml, build.zig and build.zig.zon file.

As I said above, I'm not sure why Github even still shows changes for the .h files, since I updated the package branch yesterday, but the .h file changes that are still shown look like the PR is against an outdated version of those files...

Maybe rebasing again on top `package` would fix that?

**Comment by kassane** [comment]
> Maybe rebasing again on top `package` would fix that?

Done!

File: build.zig (hunk 7 above)
[the remark this answers is not in the document]
Comment (floooh) [reply build.zig:134]: Oki doki, suppressing the output also works. Maybe I'll tinker with this a bit after merging.

**Comment by floooh** [comment]
Interesting, the captureStdOut trick doesn't seem to hide the emsdk output. But anyway, I'm happy with the PR as is and will merge now, and then start fixing the remaining issues (black screen samples etc...) in a branch wasm_zig-0.12.0

**Comment by floooh** [comment]
Ok merged! Slight correction: target branch name is https://github.com/floooh/sokol-zig/tree/zig-0.12.0-wasm

...I'll continue working on this until it's ready for merging into master and will keep you posted (I'll probably open a new PR where we can discuss changes if needed).

**Comment by floooh** [comment]
PS: Many thanks for this awesome PR :)

**Comment by kassane** [comment]
# More fixes
- https://github.com/floooh/sokol-zig/pull/51
